metals/src/main/scala/scala/meta/internal/metals/mcp/QueryEngine.scala

```diff
@@ -0,0 +1,442 @@
+            SymbolType.Class =>
+          true
+        case _ => false
+      }
+    } yield for {
+      completions <-
+        if (shouldGetCompletions) getCompletions(symbol, buildTarget)
+        else Future.successful(Nil)
+      signatures <-
+        if (shouldGetSignature) getSignatures(symbol, buildTarget)
+        else Future.successful(Nil)
+    } yield {
+      val res: Option[SymbolInspectResult] = symbol.symbolType match {
+        case SymbolType.Package =>
+          Some(PackageInspectResult(symbol.path, completions))
+        case SymbolType.Class =>
+          Some(ClassInspectResult(symbol.path, completions, signatures))
+        case SymbolType.Object =>
+          Some(ObjectInspectResult(symbol.path, completions))
+        case SymbolType.Trait =>
+          Some(TraitInspectResult(symbol.path, completions))
+        case SymbolType.Method | SymbolType.Function | SymbolType.Constructor =>
+          Some(MethodInspectResult(symbol.path, signatures, symbol.symbolType))
+        case _ => None
+      }
+      res
+    }
+
+    Future.sequence(results).map(_.flatten)
+  }
+
+  private def getCompletions(
+      symbol: SymbolSearchResult,
+      buildTarget: BuildTargetIdentifier,
+  ) = {
+    def isInteresting(completion: CompletionItem): Boolean = {
+      !QueryEngine.uninterestingCompletions(completion.getLabel())
+    }
+
+    def isDeprecated(completion: CompletionItem): Boolean = {
+      Option(completion.getTags.asScala)
+        .getOrElse(Nil)
+        .contains(CompletionItemTag.Deprecated)
+    }
+
+    compilers
+      .completions(
+        buildTarget,
+        makeCompilerOffsetParams(symbol, forSignature = false),
+      )
+      .map { completionList =>
+        completionList
+          .getItems()
+          .asScala
+          .collect {
+            case completion
+                if isInteresting(completion) && !isDeprecated(completion) =>
+              completion.getLabel()
+          }
+          .toList
+      }
+  }
+
+  private def getSignatures(
+      symbol: SymbolSearchResult,
+      buildTarget: BuildTargetIdentifier,
+  ): Future[List[String]] = {
+    compilers
+      .signatureHelp(
+        buildTarget,
+        makeCompilerOffsetParams(symbol, forSignature = true),
+      )
+      .map {
+        _.getSignatures().asScala.map(_.getLabel()).toList
+      }
+  }
+
+  private def makeCompilerOffsetParams(
+      symbol: SymbolSearchResult,
+      forSignature: Boolean,
+  ) = {
+    val lastTypeIndx =
+      (if (forSignature) symbol.symbol.stripSuffix("#") else symbol.symbol)
+        .lastIndexOf('#')
+    val completionText =
+      if (lastTypeIndx == -1) { symbol.path }
+      else {
+        val memberPart =
+          if (symbol.symbol.length() == lastTypeIndx + 1) ""
+          else "." ++ symbol.path.substring(lastTypeIndx + 1)
+        val classMemberPart = symbol.symbol
+          .substring(0, lastTypeIndx)
+          .replace('/', '.')
+          .replace('$', '.')
```

Comment:
```suggestion
          .replace('$', '.type')
```
?
`asInstanceOf` will not work for an object, no?

Reply:
Actually for:
```Scala
class Bar {
  object Foo {
    class Nested {
      def someMethod: Int = 42
    }
  }
}

???.asInstanceOf[demo.Bar#Foo.type#Nested] // won't compile

// possible working solution:
val x = ???.asInstanceOf[demo.Bar].Foo
???.asInstanceOf[x.Nested]
```